expression/builtin_string_vec.go

```diff
@@ -530,11 +530,79 @@ func (b *builtinInsertBinarySig) vecEvalString(input *chunk.Chunk, result *chunk
 }
 
 func (b *builtinConcatWSSig) vectorized() bool {
-	return false
+	return true
 }
 
+// vecEvalString evals a CONCAT_WS(separator,str1,str2,...).
+// See https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_concat-ws
 func (b *builtinConcatWSSig) vecEvalString(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	argsLen := len(b.args)
+
+	bufs := make([]*chunk.Column, argsLen)
+	var err error
+	for i := 0; i < argsLen; i++ {
+		bufs[i], err = b.bufAllocator.get(types.ETString, n)
+		if err != nil {
+			return err
+		}
+		defer b.bufAllocator.put(bufs[i])
+		if err := b.args[i].VecEvalString(b.ctx, input, bufs[i]); err != nil {
+			return err
+		}
+	}
+
+	isNulls := make([]bool, n)
+	seps := make([]string, n)
+	strs := make([][]string, n)
+	for i := 0; i < n; i++ {
+		if bufs[0].IsNull(i) {
+			// If the separator is NULL, the result is NULL.
+			isNulls[i] = true
+			continue
+		}
+		isNulls[i] = false
+		seps[i] = bufs[0].GetString(i)
+		strs[i] = make([]string, 0, argsLen-1)
+	}
+
+	var strBuf string
+	targetLengths := make([]int, n)
+	for j := 1; j < argsLen; j++ {
+		for i := 0; i < n; i++ {
+			if isNulls[i] || bufs[j].IsNull(i) {
+				// CONCAT_WS() does not skip empty strings. However,
+				// it does skip any NULL values after the separator argument.
+				continue
+			}
+			strBuf = bufs[j].GetString(i)
+			targetLengths[i] += len(strBuf)
+			if i > 1 {
+				targetLengths[i] += len(seps[i])
+			}
+			if uint64(targetLengths[i]) > b.maxAllowedPacket {
+				b.ctx.GetSessionVars().StmtCtx.AppendWarning(errWarnAllowedPacketOverflowed.GenWithStackByArgs("concat_ws", b.maxAllowedPacket))
+				isNulls[i] = true
+				continue
+			}
+			strs[i] = append(strs[i], strBuf)
+		}
+	}
+	result.ReserveString(n)
+	for i := 0; i < n; i++ {
+		if isNulls[i] {
+			result.AppendNull()
+			continue
+		}
+		str := strings.Join(strs[i], seps[i])
+		// check whether the length of result is larger than Flen
+		if len(str) > b.tp.Flen {
+			result.AppendNull()
+			continue
+		}
+		result.AppendString(str)
```

Comment:
How about skipping `strings.Join` and using a loop to append strs and sep into result directly?
`strings.Join` will result in memory copy from `strs` to `str`.

Reply:
Appending strs and sep into result has the same cost of copying.

Reply:
I will try it.